File: workers/research/src/black_book_research/relevance/override.py
"""Manual relevance override validation (BB-040)."""

from __future__ import annotations

from .types import RelevanceDecision, RelevanceOverride

MIN_OVERRIDE_REASON_LENGTH = 12


def assert_override_reason_present(reason: str) -> None:
    if len(reason.strip()) < MIN_OVERRIDE_REASON_LENGTH:
        raise ValueError(
            f"Relevance override reason is required and must be at least "
            f"{MIN_OVERRIDE_REASON_LENGTH} characters."
        )


def validate_relevance_override(
    *,
    decision: RelevanceDecision,
    reason: str,
    overridden_by: str,
    overridden_at: str,
) -> RelevanceOverride:
    if not overridden_by.strip():
        raise ValueError("Relevance override requires overridden_by.")
    if not overridden_at.strip():
        raise ValueError("Relevance override requires overridden_at.")
    assert_override_reason_present(reason)
    return RelevanceOverride(
        decision=decision,
        reason=reason.strip(),
        overridden_by=overridden_by.strip(),
        overridden_at=overridden_at.strip(),
    )

File: packages/domain/src/relevance/override.ts
/**
 * Manual relevance override with required human reason.
 */
import type { RelevanceDecision } from '@repo/schemas';
import type { RelevanceOverride } from './types.js';

const MIN_OVERRIDE_REASON_LENGTH = 12;

export type ValidateOverrideInput = {
  readonly decision: RelevanceDecision;
  readonly reason: string;
  readonly overriddenBy: string;
  readonly overriddenAt: string;
};

export function assertOverrideReasonPresent(reason: string): void {
  const trimmed = reason.trim();
  if (trimmed.length < MIN_OVERRIDE_REASON_LENGTH) {
    throw new Error(
      `Relevance override reason is required and must be at least ${MIN_OVERRIDE_REASON_LENGTH} characters.`,
    );
  }
}

export function validateRelevanceOverride(input: ValidateOverrideInput): RelevanceOverride {
  if (!input.overriddenBy.trim()) {
    throw new Error('Relevance override requires overriddenBy.');
  }
  if (!input.overriddenAt.trim()) {
    throw new Error('Relevance override requires overriddenAt.');
  }
  assertOverrideReasonPresent(input.reason);
  return {
    decision: input.decision,
    reason: input.reason.trim(),
    overriddenBy: input.overriddenBy.trim(),
    overriddenAt: input.overriddenAt.trim(),
  };
}

export function applyRelevanceOverride(
  baseDecision: RelevanceDecision,
  override?: RelevanceOverride,
): RelevanceDecision {
  return override?.decision ?? baseDecision;
}
